core/models/project.py

```diff
@@ -306,6 +318,62 @@ def publish(self):
             self.elu_accession = generate_id_function(self)
             self.save()
 
+    @staticmethod
+    def get_notification_recipients():
+        """
+        Get distinct users that are local custodian of a project.
+        """
+
+        return get_user_model().objects.filter(Q(project_set__isnull=False)).distinct()
+
+    @classmethod
+    def make_notifications(cls, exec_date: datetime.date):
+        recipients = cls.get_notification_recipients()
+        for user in recipients:
+            notification_setting: NotificationSetting = (
+                Project.get_notification_setting(user)
+            )
+            if not (
+                notification_setting.send_email or notification_setting.send_in_app
+            ):
+                continue
+            day_offset = timedelta(days=notification_setting.notification_offset)
+
+            for project in user.project_set.all():
+                # Project start date
+                if project.start_date and project.start_date - day_offset == exec_date:
+                    cls.notify(user, project, NotificationVerb.start)
+                # Project end date
+                if project.end_date and project.end_date - day_offset == exec_date:
+                    cls.notify(user, project, NotificationVerb.end)
+
```

Comment:
somehow I feel there is too much generic code in this method and that only this part should actually be at this entity level.

Reply:
I had the same feeling but the recipients are different based on the entity and also the way that inheritance works for each entity. Please suggest if you have something better in mind, I am all for it

Reply:
something like that could work?

moving the make_notifications to the mixin.

`@classmethod
    def make_notifications(cls, exec_date: "date"):
        """
        Creates a notifications for the reciepients based on
        the business logic of the entity.

        Params:
            exec_date: The date of execution

        """
        recipients = cls.get_notification_recipients()
        for user in recipients:
            notification_setting: NotificationSetting = cls.get_notification_setting(
                user
            )
            if not (
                notification_setting.send_email or notification_setting.send_in_app
            ):
                continue
            day_offset = timedelta(days=notification_setting.notification_offset)
            cls.make_notifications_for_user(day_offset, exec_date, user)`

adding an abstract make_notifications_for_user to the mixin.
and the actual implementation in the different models.

Reply:
I addressed this in the latest comment, thanks for the suggestion